File: Leetcode/663/663.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
const MAXN = 1e4 + 5
var dp map[*TreeNode]int

func solve(root *TreeNode) int {
	val, ok := dp[root]
	if ok {
		return val
	}
	dp[root] = root.Val
	if root.Left != nil {
		dp[root] += solve(root.Left)
	}
	if root.Right != nil {
		dp[root] += solve(root.Right)
	}
	return dp[root]
}

func checkEqualTree(root *TreeNode) bool {
	dp = nil
	dp = make(map[*TreeNode]int)
	// bfs
	queue := make(chan *TreeNode, MAXN)
	sumOfTree := solve(root)
	queue <- root
	for len(queue) > 0 {
		node := <- queue
		if node != root && sumOfTree == solve(node)<<1 {
			return true
		}
		if node.Left != nil {
			queue <- node.Left
		}
		if node.Right != nil {
			queue <- node.Right
		}
	}
	return false
}

File: Leetcode/663/663.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*,int> dp;
    int sumOfTree;
    bool haveAns;
    int solve(TreeNode* node) {
        if (dp.find(node) != dp.end()) {
            return dp[node];
        }
        dp[node] = node->val;
        if (node->left) {
            dp[node] += solve(node->left);
        }
        if (node->right) {
            dp[node] += solve(node->right);
        }
        return dp[node];
    }
    void dfs(TreeNode* node, TreeNode* fa) {
        if (haveAns) {
            return;
        }
        if (node) {
            if (fa && solve(node) == sumOfTree - solve(node)) {
                haveAns = true;
                return;
            }
            dfs(node->left, node);
            dfs(node->right, node);
        }
    }
    bool checkEqualTree(TreeNode* root) {
        dp.clear();
        haveAns = false;
        sumOfTree = solve(root);
        dfs(root, nullptr);
        return haveAns;
    }
};